File: src/main/java/com/ff0044/engine/graph/ShaderProgram.java
package com.ff0044.engine.graph;

import org.lwjgl.opengl.GL30;
import com.ff0044.engine.Utils;
import org.tinylog.Logger;

import javax.management.RuntimeErrorException;
import java.util.*;

import static org.lwjgl.opengl.GL30.*;

public class ShaderProgram {
    private final int programId;

    public ShaderProgram(List<ShaderModuleData> shaderModuleDataList) {
        programId = glCreateProgram();
        if (programId==0) {
            Logger.error("Could not create shader", new RuntimeException());
            throw new RuntimeException();
        }

        List<Integer> shaderModules = new ArrayList<>();
        shaderModuleDataList.forEach(s -> shaderModules.add(createShader(Utils.readFile(s.shaderFile), s.shaderType)));
        link(shaderModules);
    }

    public void bind() {
        glUseProgram(programId);
    }

    public void cleanup() {
        unbind();
        if (programId == 0) {
            glDeleteProgram(programId);
        }
    }

    protected int createShader(String shaderCode, int shaderType) {
        int shaderId = glCreateShader(shaderType);
        if (shaderId == 0) {
            Logger.error("Error creating shader. Type: " + shaderType, new RuntimeException());
            throw new RuntimeException();
        }

        // Log the shader code for debugging
        Logger.debug("Compiling shader code:\n" + shaderCode);

        glShaderSource(shaderId, shaderCode);
        glCompileShader(shaderId);

        if (glGetShaderi(shaderId, GL_COMPILE_STATUS) == 0) {
            Logger.error("Error compiling Shader code: " + glGetShaderInfoLog(shaderId, 1024), new RuntimeException());
            throw new RuntimeException();
        }

        glAttachShader(programId, shaderId);

        return shaderId;
    }

    public int getProgramId() {
        return programId;
    }

    private void link(List<Integer> shaderModules) {
        glLinkProgram(programId);
        if (glGetProgrami(programId, GL_LINK_STATUS) == 0) {
            Logger.error("Error linking shader code: " + glGetProgramInfoLog(programId, 1024), new RuntimeException());
            throw new RuntimeException();
        }

        shaderModules.forEach(s->glDetachShader(programId, s));
        shaderModules.forEach(GL30::glDeleteShader);
    }

    public void unbind() {
        glUseProgram(0);
    }

    public void validate() {
        glValidateProgram(programId);
        if (glGetProgrami(programId, GL_VALIDATE_STATUS) == 0) {
            Logger.error("Error validating shader code: " + glGetProgramInfoLog(programId, 1024), new RuntimeException());
            throw new RuntimeException();
        }
    }

    public record ShaderModuleData(String shaderFile, int shaderType) {}
}

File: src/main/kotlin/tk/ff0044/lwkglTutorial/engine/graph/ShaderProgram.kt
package tk.ff0044.lwkglTutorial.engine.graph

import org.joml.Matrix4f
import org.lwjgl.opengl.GL20.*
import org.lwjgl.system.MemoryStack
import org.tinylog.Logger

class ShaderProgram {
    private var programId : Int = 0
    private var vertexShaderId = 0
    private var fragmentShaderId = 0
    private var uniforms: MutableMap<String, Int> = HashMap()

    init {
        programId = glCreateProgram()
        if (programId == 0) {
            Logger.error { "Could not create Shader" }
            throw Exception("Could not create Shader")
        }

        val error = glGetError()
        if (error != GL_NO_ERROR) {
            Logger.error { "OpenGL error: $error" }
            throw Exception("OpenGL error: $error")
        }
    }

    @Throws(Exception::class)
    fun createUniform(uniformName: String) {
        val uniformLocation = glGetUniformLocation(programId, uniformName)
        if (uniformLocation < 0) {
            throw Exception("Could not find uniform: $uniformName")
        }
        uniforms[uniformName] = uniformLocation
    }

    @Throws(Exception::class)
    fun setUniform(uniformName: String, value: Matrix4f) {
        // Dump the matrix into a float buffer
        MemoryStack.stackPush().use { stack ->
            glUniformMatrix4fv(uniforms[uniformName]!!, false, value[stack.mallocFloat(16)])
        }
    }

    @Throws(Exception::class)
    fun createVertexShader(shaderCode: String) {
        vertexShaderId = createShader(shaderCode, GL_VERTEX_SHADER)
    }

    @Throws(Exception::class)
    fun createFragmentShader(shaderCode: String) {
        fragmentShaderId = createShader(shaderCode, GL_FRAGMENT_SHADER)
    }

    @Throws(Exception::class)
    private fun createShader(shaderCode: String, shaderType: Int): Int {
        val shaderId = glCreateShader(shaderType)
        if (shaderId == 0) {
            throw Exception("Error creating shader. Type: $shaderType")
        }

        glShaderSource(shaderId, shaderCode)
        glCompileShader(shaderId)

        if (glGetShaderi(shaderId, GL_COMPILE_STATUS) == 0) {
            val infoLog = glGetShaderInfoLog(shaderId, 1024)
            glDeleteShader(shaderId)
            throw Exception("Error compiling Shader code: $infoLog")
        }

        glAttachShader(programId, shaderId)
        return shaderId
    }

    @Throws(Exception::class)
    fun link() {
        glLinkProgram(programId)
        if (glGetProgrami(programId, GL_LINK_STATUS) == 0) {
            val infoLog = glGetProgramInfoLog(programId, 1024)
            throw Exception("Error linking Shader code: $infoLog")
        }

        if (vertexShaderId != 0) {
            glDetachShader(programId, vertexShaderId)
        }
        if (fragmentShaderId != 0) {
            glDetachShader(programId, fragmentShaderId)
        }

        glValidateProgram(programId)
        if (glGetProgrami(programId, GL_VALIDATE_STATUS) == 0) {
            Logger.warn{"Warning validating Shader code: " + glGetProgramInfoLog(programId, 1024)}
        }
    }

    fun bind() {
        glUseProgram(programId)
    }

    fun unbind() {
        glUseProgram(0)
    }

    fun cleanup() : Int {
        unbind()
        Logger.debug{"Shader program has been unbinded"}
        if (programId != 0) {
            glDeleteProgram(programId)
            Logger.debug{"Shader program has been deleted"}
        }
        return 1
    }
}